Adding `Avx10v1` to the runtime

Ruihan-Yin opened the pull request:
# Overview

As discussed in #98069, we will expose `Avx10` converged vector ISA into the runtime. This PR is adding CPUID detection logics of the first version. 

Following the [spec doc](https://cdrdv2-public.intel.com/784267/355989-intel-avx10-spec.pdf) (C1 1.2), availability of `Avx10v1` will be checked in the given paradigm:
1. A CPUID feature bit indicating that the Intel AVX10 ISA is supported.
2. A version number to ensure that the supported version is greater than or equal to the desired version (say 1, in this case).
3. A vector length bit indicating the maximum supported vector length.

We also exposed the following environment variables to control the ISA support:
1. `Avx10MaxVectorLength`: an integer that specifies the max vector length expected.
2. `EnableAvx10v1`: covers all the instructions operating on the length of 128.
3. `EnableAvx10v1_V256`: covers all the instructions operating on the length of 256.
4. `EnableAvx10v1_V512`: covers all the instructions operating on the length of 512.

`Avx10MaxVectorLength` has priority over the rest of the three vars, e.g. when `Avx10MaxVectorLength < 512` and `EnableAvx10v1_V512 = 1`, Instructions under `Avx10v1_V512` will not be available.

Ruihan-Yin committed:
script-gen changes

Ruihan-Yin committed:
hand-written changes

DeepakRajendrakumaran commented:
@dotnet/avx512-contrib @BruceForstall Is there a similar group for 'arch-avx10' that we can add? Don't think we have permission to add the 'arch-avx10' label

tannergooding commented on src/coreclr/inc/clrconfigvalues.h:
```diff
@@ -760,6 +760,10 @@ RETAIL_CONFIG_DWORD_INFO(EXTERNAL_EnableAVX512F,                W("EnableAVX512F
 RETAIL_CONFIG_DWORD_INFO(EXTERNAL_EnableAVX512VBMI,             W("EnableAVX512VBMI"),          1, "Allows AVX512VBMI+ hardware intrinsics to be disabled")
 RETAIL_CONFIG_DWORD_INFO(EXTERNAL_EnableAVX512VBMI_VL,          W("EnableAVX512VBMI_VL"),       1, "Allows AVX512VBMI_VL+ hardware intrinsics to be disabled")
+RETAIL_CONFIG_DWORD_INFO(EXTERNAL_EnableAVX10v1,                W("EnableAVX10v1"),             1, "Allows AVX10v1+ hardware intrinsics to be disabled")
+RETAIL_CONFIG_DWORD_INFO(EXTERNAL_EnableAVX10v1_V256,           W("EnableAVX10v1_V256"),        1, "Allows AVX10v1_V256+ hardware intrinsics to be disabled")
+RETAIL_CONFIG_DWORD_INFO(EXTERNAL_EnableAVX10v1_V512,           W("EnableAVX10v1_V512"),        1, "Allows AVX10v1_V512+ hardware intrinsics to be disabled")
+RETAIL_CONFIG_DWORD_INFO_EX(EXTERNAL_Avx10MaxVectorLength,      W("Avx10MaxVectorLength"),      0, "The max vector length supported", CLRConfig::LookupOptions::ParseIntegerAsBase10)
```
I don't think we need these three.

In general these `DOTNET_Enable{Isa}` knobs are primarily there for testing purposes. That is, they exist so that developers with newer hardware can still test downlevel code paths without having to recompile their applications. Such code is not typically meant for use in actual production scenarios.

To that end, we typically expose one knob per CPUID ISA bit and so you can disable `Avx` but not `Avx.X64`. `Avx512F.VL` was somewhat an exception and in hindsight, probably a knob we don't actually need to let users control. I expect that the `V256` and `V512` nested classes fit into the same bucket, where users don't need the ability to enable/disable them directly and independently of `Avx10` itself.

Instead, users who need to control access to the used vector bit width can utilize the `DOTNET_PreferredVectorBitWidth` knob and the corresponding `Vector###.IsHardwareAccelerated` checks, which is how they're expected to detect this in existing code paths.

tannergooding replied:
-- @BruceForstall, there should be no issue in removing a config knob like the various `DOTNET_EnableAvx512*_VL` entries, right?

We can simply remove them and document that users should utilize `DOTNET_PreferredVectorBitWidth` and `DOTNET_EnableAvx512*` instead? This would allow a consistent user story here and reduce the overall complexity we need to support/consider

BruceForstall replied:
I don't see any problem with removing the `DOTNET_EnableAvx512*_VL` configs. They are available in Release and have been shipped, and use the `EXTERNAL` prefix, which indicates (or at least historically indicated) a documented config. (Maybe we should have used the prefix `UNSUPPORTED`?) So technically removing them is a breaking change, but it doesn't seem like a problem.

Ruihan-Yin replied:
Thanks for the feedback.

Please correct me if I am wrong, so we want remove `DOTNET_EnableAvx10v1_V256/512` and `DOTNET_Avx10MaxVectorLength`. And keep `DOTNET_EnableAvx10v1`, let it plus `DOTNET_PreferredVectorBitWidth` to control the emit behavior of `Vector256/512` APIs. (Which I suppose won't be covered in this PR.)

As for the `VL` vars, do we want to handle it in this PR, or separately?

tannergooding replied:
> so we want to

Yep!

> Which I suppose won't be covered in this PR

This should already be somewhat implicit based on the existing checks we have in the JIT, so I don't expect we need to do anything special. Users will still see `Avx10v1.V512` report supported if the hardware actually supports it and they would simply check `Vector512.IsHardwareAccelerated` if they don't want limit 512-bit usage to the user selected behavior (which is controlled by `PreferredVectorBitWidth`).

> As for the VL vars, do we want to handle it in this PR, or separately?

I'll cover it in a separate PR.

tannergooding commented on src/coreclr/tools/Common/JitInterface/ThunkGenerator/InstructionSetDesc.txt:
```diff
@@ -57,6 +57,9 @@ instructionset     ,X86   ,Avx512DQ             ,        ,35 ,AVX512DQ
 instructionset     ,X86   ,Avx512DQ_VL          ,        ,36 ,AVX512DQ_VL           ,avx512dq_vl
 instructionset     ,X86   ,Avx512Vbmi           ,        ,37 ,AVX512VBMI            ,avx512vbmi
 instructionset     ,X86   ,Avx512Vbmi_VL        ,        ,38 ,AVX512VBMI_VL         ,avx512vbmi_vl
+instructionset     ,X86   ,Avx10v1              ,        ,44 ,AVX10v1               ,avx10v1
+instructionset     ,X86   ,Avx10v1_V256         ,        ,45 ,AVX10v1_V256          ,avx10v1_v256
+instructionset     ,X86   ,Avx10v1_V512         ,        ,46 ,AVX10v1_V512          ,avx10v1_v512
```
We're missing the `instructionset64bit,X86 ,*` entries for these three.

We're also missing the `impliciation` entries (`Avx10v1` implies at least `Avx2` and `FMA`; `Avx10v1_V512` implies most or all of the `Avx512` ISAs)

Ruihan-Yin replied:
trying to make a list of the implication here:
`Avx10v1` implies:
1. Avx2
2. FMA

`Avx10v1_V512` implies: (From [comment](https://github.com/dotnet/runtime/issues/98069#issue-2121768651) in #98069 )
1.  Avx512F
2. Avx512BW
3. Avx512DQ
4. Avx512CD
5. Avx512_VNNI
6. Avx512_VBMI


(There are most on the original list, but here I listed the intersection part with what has been defined in InstructionSetDesc.txt )

Please correct me if I missed anything here.

tannergooding replied:
Looking at the docs, its `F`, `CD`, `BW`, `DQ`, `VBMI`, `IFMA`, `VNNI`, `BF16`, `VPOPCNTDQ`, `VBMI2`, `BITALG`, and `FP16`.

It's not listed, but I also believe `Avx10v1_V512` implies the `_VL` subsets (since those are just 128/256-bit variants which are re-exposed by `Avx10v1` and `Avx10v1_V256`).

There is also `VAES`, `FGNI`, and `VPCLMULQDQ` but they continue to be iterated independently and are not implied.

![image](https://github.com/dotnet/runtime/assets/10487869/c3eae176-1a47-47ac-ae53-ccec2ac790ae)

tannergooding commented on src/native/minipal/cpufeatures.c:
```diff
@@ -306,6 +306,42 @@ int minipal_getcpufeatures(void)
+                const int versionMask = 0xFF; // [7:0]
+                if((cpuidInfo[CPUID_EBX] & versionMask) >= 1) // version higher than 1
+                {
+                    result |= XArchIntrinsicConstants_VectorT128; // Avx10 guarantee V128
```
This shouldn't be required, `Avx10` requires `Avx2` and so `VectorT128` and `VectorT256` should already be set.

Similarly, since `Avx10/512` implies `Avx512F`, `VectorT512` will also have already been set above.

tannergooding commented on src/native/minipal/cpufeatures.c:
```diff
@@ -306,6 +306,42 @@ int minipal_getcpufeatures(void)
+            // with 1,2, set XArchIntrinsicConstants_Avx10_1
+
+            // 3. CPUID.(EAX=24H, ECX=00H):EBX[bits 18:16] -> vector length support.
+            __cpuidex(cpuidInfo, 0x00000007, 0x00000001);
```
I think we can put this further up, where we already query `__cpuidex(cpuidInfo, 0x00000007, 0x00000001)` as part of `AvxVnni`

That path is after the `Avx2` check and so its only happening after we've already confirmed other required ISAs are supported.

tannergooding commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -1448,6 +1448,21 @@ void EEJitManager::SetCpuInfo()
+        CPUCompileFlags.Set(InstructionSet_AVX10v1);
+    }
+
+    if (((cpuFeatures & XArchIntrinsicConstants_Avx10v1_V256) != 0) && CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_EnableAVX10v1))
```
I don't think we need this `CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_EnableAVX10v1)` check.

The `EnsureValidInstructionSetSupport` call below should ensure it is correct provided the `implication` set in `InstructionSetDesc.txt` are correct

Ruihan-Yin replied:
you mean here we only check if XArchIntrinsicConstants are set for Avx10v1_V256/512 and make `Avx10v1_V512` implies `Avx10v1_V256` and `Avx10v1_V256` implies `Avx10v1`?

This is confirmed in Avx10 spec doc:
>Developers can assume that the highest supported vector length for a processor implies that all lesser vector
lengths are also supported. Scalar Intel AVX-512 instructions will be supported independent of the maximum
vector width.

tannergooding replied:
Correct. Once the implications are correct, then `EnusreValidInstructionSetSupport` should let everything else flow out, so we only need to check for `CLRConfig::EXTERNAL_EnableAVX10v1` the once

tannergooding commented on src/coreclr/tools/Common/Internal/Runtime/ReadyToRunInstructionSet.cs:
```diff
@@ -55,6 +55,9 @@ public enum ReadyToRunInstructionSet
         VectorT512=41,
         Rcpc2=42,
         Sve=43,
+        Avx10v1=44,
```
I think we're missing some changes to
* `src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs`
* `src/coreclr/tools/Common/InstructionSetHelpers.cs`

When the empty classes get add to the managed side, we'd then have a few other places to update as well; but my guess is you're planning on those going into a separate PR?

Ruihan-Yin replied:
I missed that part, those should also be covered by this PR. Thanks for pointing out.

Ruihan-Yin committed:
Add missing definition and implication in ISA descriptor.

Ruihan-Yin committed:
improve the CPUID check logic.

Ruihan-Yin committed:
Add missing changes in managed code and fix bad naming.

Ruihan-Yin committed:
Resolve comments:
Removed seperate env vars to control Avx10v1_V256/512, now Avx10v1 instructions with different vector length will all be controlled by EnableAvx10v1 alone.

tannergooding commented on src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs:
```diff
@@ -69,6 +69,9 @@ private static class XArchIntrinsicConstants
             public const int VectorT128 = 0x4000000;
             public const int VectorT256 = 0x8000000;
             public const int VectorT512 = 0x10000000;
+            public const int Avx10v1 = 20000000;
+            public const int Avx10v1_v256 = 40000000;
+            public const int Avx10v1_v512 = 50000000;
```
nit: These should be `hex` and the last one should be `8`, not `5`.

Ruihan-Yin replied:
`public const int Avx10v1_v512 = 0x80000000;`

This line will raise an error that 0x80000000 cannot be implicitly converted to `int`, although it can be fixed by explicitly defining as:
`public const int Avx10v1_v512 = -2147483648;`, I wonder if this is a proper fix (I suppose it should be good in terms of the number itself as it is just a flag bit), considering it has reached the max number of ISA this design can hold, and we are expecting some more ISAs coming in the future, e.g. higher version of Avx10 and APX, etc.

tannergooding replied:
Either `public const int Avx10v1_v512 = -2147483648;` or `public const int Avx10v1_v512 = unchecked((int)0x80000000));` should be fine

Changing it to `uint` or `ulong` to hold more bits might also be appropriate, it's probably a better question for @MichalStrehovsky on how he wants it handled long term for this part of the codebase.

MichalStrehovsky replied:
> Changing it to uint or ulong to hold more bits might also be appropriate, it's probably a better question for @MichalStrehovsky on how he wants it handled long term for this part of the codebase.

I don't have an opinion; whatever works. At this pace we may run out of ulong bits too. I don't have a good understanding for why we need so many bits: the VL suffixes look redundant; the various VectorT are also redundant with other flags we set at the same time, etc.

I was involved in the design for this up to AVX2/BMI but haven't really though about it since. The old design doesn't scale if we use bits at this pace because we can't make SSE3/SSE41/etc. baseline yet (and free up bits that way), but we may already need new bits. It would be better for someone who actually understands this to design a shape that will work.

There is an advantage if this can fit in 32bits (run-time check when IsSupported is done dynamically at runtime use the same enum and 32bits are simpler), but we can also introduce yet another enum for those purposes that doesn't waste bits, or have two ints, or something like that. If we can reclaim those 7 bits, it might be enough breathing room to keep using this enum and 32bits.

tannergooding commented on src/coreclr/tools/Common/InstructionSetHelpers.cs:
```diff
@@ -224,6 +224,30 @@ internal static partial class Helpers
                     optimisticInstructionSetSupportBuilder.AddSupportedInstructionSet("avx512vbmi_vl");
                 }
+
+                Debug.Assert(InstructionSet.X64_AVX10v1_V256 == InstructionSet.X86_AVX10v1_V256);
+                if (supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX10v1_V256))
+                {
+                    Debug.Assert(supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX10v1));
+                    Debug.Assert(supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX2));
+                    Debug.Assert(supportedInstructionSet.HasInstructionSet(InstructionSet.X64_FMA));
+                }
+
+                Debug.Assert(InstructionSet.X64_AVX10v1_V512 == InstructionSet.X86_AVX10v1_V512);
+                if (supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX10v1_V512))
+                {
+                    Debug.Assert(supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX10v1_V256));
+                    Debug.Assert(supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX512F));
+                    Debug.Assert(supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX512F_VL));
+                    Debug.Assert(supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX512BW));
+                    Debug.Assert(supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX512BW_VL));
+                    Debug.Assert(supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX512CD));
+                    Debug.Assert(supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX512CD_VL));
+                    Debug.Assert(supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX512DQ));
+                    Debug.Assert(supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX512DQ_VL));
+                    Debug.Assert(supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX512VBMI));
+                    Debug.Assert(supportedInstructionSet.HasInstructionSet(InstructionSet.X64_AVX512VBMI_VL));
+                }
```
I'm not sure we need these.

The main intent of these paths is to add `optimisticInstructionSet` that is ones where a cached CPUID check is beneficial for AOT code. I don't think that's worth doing for `Avx10v1` here since it would only be applicable if `Avx512F` was there and that itself requires `VL` support.

The `Debug.Assert` where then there to setup the expectation that `Avx512F` was only supported when `BW+CD+DQ+VL` were also supported, even though they are technically supersets of `Avx512F`. We don't have a similar consideration here so no `Debug.Assert` is needed

tannergooding commented on src/native/minipal/cpufeatures.c:
```diff
@@ -277,6 +277,28 @@ int minipal_getcpufeatures(void)
+                                                const int versionMask = 0xFF; // [7:0]
+                                                if((cpuidInfo[CPUID_EBX] & versionMask) >= 1) // version higher than 1
+                                                {
+                                                    result |= XArchIntrinsicConstants_Avx10v1;
```
nit: We should explicitly check `EBX[16]` is set to indicate `AVX10/V128`. The bit exists, so we should be explicit about checking it.

tannergooding commented on src/native/minipal/cpufeatures.c:
```diff
@@ -277,6 +277,28 @@ int minipal_getcpufeatures(void)
+                                                    result |= XArchIntrinsicConstants_Avx10v1;
+                                                }
+
+                                                const int vector256Mask = (1 << 17);
+                                                const int vector512Mask = (1 << 18);
+                                                if((cpuidInfo[CPUID_EBX] & vector256Mask) != 0)
+                                                {
+                                                    result |= XArchIntrinsicConstants_Avx10v1_V256;
+                                                }
+
+                                                if((cpuidInfo[CPUID_EBX] & vector512Mask) != 0)
+                                                {
+                                                    result |= XArchIntrinsicConstants_Avx10v1_V512;
+                                                }
```
nit: These can only be considered if `AVX10/V128` is set and if the `version` was greater than `1`, so we should have them appropriately nested

tannergooding commented on src/native/minipal/cpufeatures.c:
```diff
@@ -277,6 +277,28 @@ int minipal_getcpufeatures(void)
+                                                    result |= XArchIntrinsicConstants_Avx10v1;
+                                                }
+
+                                                const int vector256Mask = (1 << 17);
```
nit: None of the other checks use a `const int` like this, they inline the `(1 << 17)` into the if statement and then add a comment indicating what bit it was checking

Ruihan-Yin replied:
I can make it to the commonly used form, it looks unnecessary to define this way as the bit-shift form should be easy enough to read in this case.

Ruihan-Yin committed:
Resolve comments:
1. Make sure the version check and length check are correctly nested.
2. remove unnecessary debug assert in mananged code.

tannergooding reviewed (approved):
LGTM.

tannergooding commented:
CC. @dotnet/jit-contrib  for secondary review and merge

BruceForstall commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -1448,6 +1448,21 @@ void EEJitManager::SetCpuInfo()
+        CPUCompileFlags.Set(InstructionSet_AVX10v1);
+    }
+
+    if (((cpuFeatures & XArchIntrinsicConstants_Avx10v1_V256) != 0))
```
Doesn't this also need `&& CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_EnableAVX10v1)`?

And the Avx10v1_V512 case below? 

That is, shouldn't all the AVX10 instruction sets be subject to the same configuration flag? (Since you didn't add separate flags for each)

Ruihan-Yin replied:
Please see the [discussion](https://github.com/dotnet/runtime/pull/99784#discussion_r1526819552) here.

We intended to let the implication that `Avx10v1_V512` guarantees `Avx10v1_V256`, and `Avx10v1_V256` guarantees `Avx10v1` to handle the invalid case (The implication is confirmed by the Avx10 spec doc.). Say, `EnusreValidInstructionSetSupport ` would cancel the illegal flag combination set during the flow.

BruceForstall replied:
Comments would be useful.

BruceForstall replied:
So on an AVX10 + V512 machine, with `DOTNET_EnableAVX10v1=0`, we'll leave this code with:
```
InstructionSet_AVX10v1_V256
InstructionSet_AVX10v1_V512
```
set but `InstructionSet_AVX10v1` unset.

Then, we call `CPUCompileFlags.EnsureValidInstructionSetSupport();` to "clean it up" in EnsureInstructionSetFlagsAreValid` which has this new code:

```
        if (resultflags.HasInstructionSet(InstructionSet_AVX10v1_V512) && !resultflags.HasInstructionSet(InstructionSet_AVX10v1_V256))
            resultflags.RemoveInstructionSet(InstructionSet_AVX10v1_V512);
        if (resultflags.HasInstructionSet(InstructionSet_AVX10v1_V256) && !resultflags.HasInstructionSet(InstructionSet_AVX10v1))
            resultflags.RemoveInstructionSet(InstructionSet_AVX10v1_V256);
        if (resultflags.HasInstructionSet(InstructionSet_AVX10v1) && !resultflags.HasInstructionSet(InstructionSet_AVX2))
            resultflags.RemoveInstructionSet(InstructionSet_AVX10v1);
```

The first case won't hit, but the others will. That seems to mean that `InstructionSet_AVX10v1_V512` will not be disabled.

If the code ran in a different order, it would work (i.e., check from deepest in the dependency tree first).

Ruihan-Yin replied:
I see, thanks for pointing out, in that case, if we want to keep the idea that only check `EnableAvx10v1` once, I can make `Avx10v1_V512` implies `Avx10v1` and `Avx10_V256`, so there will be an extra check like:
```
        if (resultflags.HasInstructionSet(InstructionSet_AVX10v1_V512) && !resultflags.HasInstructionSet(InstructionSet_AVX10v1_V128))
            resultflags.RemoveInstructionSet(InstructionSet_AVX10v1_V512);
```
this should be able to avoid the described case. 

I presume changing the order for the checks could fix the issue but might not be a solid solution, as if the order changes somehow in the future, this "implicit" bug could be hard to find.

BruceForstall replied:
I guess it makes sense to add:
```
implication        ,X86   ,AVX10v1_V512         ,AVX10v1
```
and the auto-generated code will do the rest?

Seems like either implication handling could be improved to handle transitive implications better, or, implications are not expected to be transitive. It would be helpful if the documentation/comments made it clear.

Ruihan-Yin replied:
I agree, I will document the conclusion in the code.

tannergooding replied:
I think the issue may just be in the order the implications are listed in the doc. We typically want the newest ISAs listed at the bottom

BruceForstall reviewed (commented):
Please resolve the merge conflict with the JIT GUID

BruceForstall commented:
If this PR is no longer in "Draft", then please mark it "Ready for review".

Ruihan-Yin commented:
> Please resolve the merge conflict with the JIT GUID

I am not sure how I can resolve this conflict, shall I just rebase the branch and overwrite the GUID with the one in my branch, or I need to rebase and then re-generate the ID?

BruceForstall commented:
> I am not sure how I can resolve this conflict, shall I just rebase the branch and overwrite the GUID with the one in my branch, or I need to rebase and then re-generate the ID?

I believe "merge" is preferred over "rebase". 

In any case, simply overwrite the GUID with yours. No need to create a new GUID, since yours is already new and hasn't been used.

Ruihan-Yin committed:
Resolve comment:
1. update the ISA implication

Ruihan-Yin committed:
Merge branch 'main' into Avx103

tannergooding commented on src/coreclr/tools/Common/JitInterface/ThunkGenerator/InstructionSetDesc.txt:
```diff
@@ -138,6 +138,7 @@ implication        ,X86   ,AVX512VBMI           ,AVX512BW
+implication        ,X86   ,AVX10v1_V512         ,AVX10v1
 implication        ,X86   ,AVX10v1_V256         ,AVX10v1
 implication        ,X86   ,AVX10v1              ,AVX2
 implication        ,X86   ,AVX10v1              ,FMA
```
I'm pretty sure the logic would be fixed if we just reordered it like this:
```suggestion
implication        ,X86   ,AVX10v1              ,AVX2
implication        ,X86   ,AVX10v1              ,FMA
implication        ,X86   ,AVX10v1_V256         ,AVX10v1
implication        ,X86   ,AVX10v1_V512         ,AVX10v1_V256
```

I think the issue stemmed from putting v512 depending on v256 prior to the v256 depending on 10v1.

Ruihan-Yin replied:
so when we define `implication <arch> <ISA> <implied ISA>`, we would need to keep the new ISA at the very bottom, and does the order of implied ISAs matters when a single ISA implies a set of ISAs, like `Avx10v1_V512` here. (I presume not.

tannergooding replied:
> so when we define implication <arch> <ISA> <implied ISA>, we would need to keep the new ISA at the very bottom

I believe so, yes. In general I think we just want to make sure `implication <arch> <ISA2> <ISA1>` is defined after `implication <arch> <ISA2> <...>`

The tool that generates the implication checks is fairly naive and just does them from top to botttom, so I think the issue stems from `Avx10v1_V512` currently being at the "top" of the `Avx10v1` implications, so it gets checked first and isn't filtered out by the other checks.

> does the order of implied ISAs matters when a single ISA implies a set of ISAs, like Avx10v1_V512 here?

I don't think so, but I think it would be good (in general) to simply match the general definition order, for simplicity.

Ruihan-Yin committed:
Resolve comments:
adjust the order of ISA implication.

Ruihan-Yin commented:
Failures look irrelevant.
